### ISSUE TITLE
Rectangular artifacts using `labelisation_seeded_watershed`

### ISSUE THREAD

**Issue by garridoq** [higra/Higra#172]
Using `labelisation_seeded_watershed` causes unwanted rectangular artifacts in the final segmentation, whereas `labelisation_watershed` does not.
This bug was reproduced using version `0.4.5`.

This issue was present using the following data
Gradient:
![Gradient](https://user-images.githubusercontent.com/26287517/74341660-5f7b9300-4da8-11ea-98ef-ac439d0e5ccb.png)

Seeds:
![Seeds](https://user-images.githubusercontent.com/26287517/74341713-71f5cc80-4da8-11ea-9e6f-9d76e1fd0f49.png)

The result was:
![seeded_ws](https://user-images.githubusercontent.com/26287517/74341754-889c2380-4da8-11ea-9482-67e025b80df1.png)

The code that can be used to reproduce this issue is :
```
gradient = np.load("gradient.npy")
seeds = np.load("seeds.npy")

im_graph = hg.get_4_adjacency_graph(gradient.shape)
im_edge_weights = hg.weight_graph(im_graph, gradient, hg.WeightFunction.mean)

labels = hg.labelisation_seeded_watershed(im_graph,im_edge_weights,seeds.flatten())
```
All the relevant files are available in [data.zip](https://github.com/higra/Higra/files/4192698/data.zip)
we also added a file called `seeds_valued.npy` where seeds have various values and not just 1, this caused the same issue with our resulting labelisation.

By overlaying the seeds with our resulting labelisation, we can indeed see that certain seeds created these region, when they are expected to be part of the same object as their neighboring seeds.
![seeds_overlay](https://user-images.githubusercontent.com/26287517/74342442-c8afd600-4da9-11ea-9692-1cd4fb67c8bc.png)

**Comment by PerretB** [comment]
Thanks for the nice and complete report ! There is indeed two things going on there. 

First, there is a bug in `labelisation_seeded_watershed` in `0.4.5`. There is a patch available on master but not yet released. If you need it before the next release (which should come next week), you can simply copy past the code of the new function https://github.com/higra/Higra/blob/38bcf46acfb6d31c531ad1906a2e2d70d57338b1/higra/algo/watershed.py#L43-L102  Note that this function has a limitation: the seeds must be connected. If this constraint doesn't fit your needs, there is a workaround detailed bellow. I will probably include a better version of the algorithm that doesn't have this limitation in the next release but this requires some c++ code to be done efficiently. 

Second, in your example, you have several seeds with different labels in a same minima of the gradient. In such case any frontier which stays inside the minima is a valid separation between those seeds (the gradient doesn't give any information), and the block artifacts are the arbitrary decisions taken by the algorithm. 

The idea to overcome the "connected seeds" constraint is the following:
1. perform a connected component labeling of the seed
2. perform a seeded watershed with those connected component labels
3. remap watershed labels to original seed labels, distinct watershed regions will get the same label if the seed they contain had the same original label
This gives something like that:

```python
gradient = np.load("gradient.npy")
seeds = np.load("seeds_valued.npy")

# convert seeds to int
_, seeds = np.unique(seeds, return_inverse=True)
seeds = seeds.reshape(gradient.shape)

# 4-connected component labelling
labeled_seeds = skimage.measure.label(seeds > 0, background=0, connectivity=1)

# seeded watershed
im_graph = hg.get_4_adjacency_graph(gradient.shape)
im_edge_weights = hg.weight_graph(im_graph, gradient, hg.WeightFunction.mean)
labels = labelisation_seeded_watershed(im_graph, im_edge_weights, labeled_seeds)

#remap labels
flat_labels = labels.ravel()
flat_seeds = seeds.ravel()
seed_positions, = np.nonzero(flat_seeds)
seed_values = flat_seeds[seed_positions]
label_values = flat_labels[seed_positions]
representative_label_values, representative_label_positions = np.unique(label_values, return_index=True)
representative_seed_values = seed_values[representative_label_positions]
labels = representative_seed_values[labels - 1]
```

**Comment by garridoq** [comment]
Thanks a lot for your answer !

Having the seeds be connected is too strong of a constraint for my use case, however your workaround works perfectly with the updated `labelisation_seeded_watershed`.

**Comment by PerretB** [comment]
Fixed by #173
